### PR TITLE
AI - add check on whether allied or enemy attacks when selecting max.

### PR DESCRIPTION
**Pull request by ron-murhammer**
This was causing issues with enemy attacks being skipped that shouldn't
have been.

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/531)

### CODE REVIEW

**Comment by DanVanAtta** [comment]
tests?

**Comment by ron-murhammer** [comment]
@DanVanAtta - Here is a save game before the change that shows the USA landing planes in Soviet Far East as it doesn't see the Japanese potential attack there. After this change you'll see, the USA doesn't land its planes in that exposed location: 
[far_east_air_fast_ai.zip](https://github.com/triplea-game/triplea/files/176511/far_east_air_fast_ai.zip)

Would be nice to eventually create a suite of unit tests to verify both max potential moves and max potential counterattacks from enemies.

**Comment by DanVanAtta** [comment]
Could you make that eventually be now? It's a 30 line static method.

**Comment by DanVanAtta** [comment]
Plus, I agree it would be great to have a suite of tests, but I'm not sure why the scope needs to be increased so dramatically.  There is no ask for a suite of tests, but for this change to be tested (and not the manual test-once, but never again variety)

**Comment by ron-murhammer** [comment]
@DanVanAtta So this method only tests that it selects the strongest group of units considering turn order when given a list of them. There are also many different varieties such as whether we are checking defenders or attackers and enemies or allies. The biggest overhead is having to mock up the data as you have a list of Maps that have ProTerritory objects which contain max units. You also need to have the data mocked up to simulate territory, players, relationships, and turn order (essentially loading up a specific map). You'd probably want to test a couple different maps and a couple different combinations of attackers/defender/enemies/allies along with the actual groups of units. So I'd say you'd need a min of 10 unit tests to even come a majority of the basic scenarios.

And unit testing just that method with those scenarios doesn't even prove that it fixes the overall issue of the AI choosing to move planes to an exposed location. You'd need a higher level unit test that tests are group of methods to prove that the AI now properly sees enemy counter attacks for this given scenario (essentially the method that calls this method).

At this point, I don't really want to spend the hours writing a bunch of units tests for a 1 line fix here and I think so far we haven't been holding up or rejecting PRs based on whether or not they have unit tests.

**Comment by DanVanAtta** [comment]
We've included test in most PRs, perhaps the one PR where I blindly fixed it and ran the code you asked for a test.

No way this needs 10 test case, nor do you need multiple maps. Add the GameData object as a method parameter, and then the rest gets pretty easy to mock out. You only need to set up a territory map that has one set of units.

You could also look to breaking up the method into multiple helper methods, then test the helper methods and leave the high level aggregation of them untested.

File: src/games/strategy/triplea/ai/proAI/data/ProOtherMoveOptions.java [hunk 1]
```diff
@@ -52,6 +54,7 @@ public String toString() {
 
   private static Map<Territory, ProTerritory> createMaxMoveMap(final List<Map<Territory, ProTerritory>> moveMaps,
       final PlayerID player, final boolean isAttacker) {
+    final GameData data = ProData.getData();
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/data/ProOtherMoveOptions.java]: Why is this set on line 57 but then only used on line 75?

Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/data/ProOtherMoveOptions.java]: ProData.getData() is essentially a global AI variable. Its holding the game data for the current phase. I tend to just put the declaration at the top of the method when its being used as it acts as almost an input parameter.

**Comment by ron-murhammer** [comment]
@DanVanAtta I'd actually say most PRs haven't included unit tests. Though depends what we define as needing unit tested. Should UI related code be unit tested? Maybe only parts that include some sort of logic? As I believe most of the recent PRs related to UI code don't have unit tests.

File: src/games/strategy/triplea/ai/proAI/data/ProOtherMoveOptions.java (hunk 1 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/data/ProOtherMoveOptions.java]: Please minimize vertical distance between usage and declaration. The repeated method call is not a performance concern. Reducing the vertical distance reduces mental burden to understand this method, it's just one more thing to remember when looking at hte for loop.

**Comment by DanVanAtta** [comment]
> "Should UI related code be unit tested?"

Not relevant, we are talking about a perfectly testable piece of logic code.

If this needs 10 different unit tests, and unit tests are simpler than functional/acceptance tests, please provide the save games that test the other variety of the cases. The save game provided certainly can't cover them all.

File: src/games/strategy/triplea/ai/proAI/data/ProOtherMoveOptions.java (hunk 1 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/data/ProOtherMoveOptions.java]: Done.

**Comment by ron-murhammer** [comment]
@DanVanAtta It is relevant since you are claiming that "We've included test in most PRs" which I don't agree with and I can probably go through the last 10-20 PRs and show that.

That save game doesn't cover them all. It covers only the case that was broken. The code only didn't work properly for finding the max 'enemy' 'attack' on an 'allied' territory. This method is also used for other combinations such as 'allied' 'attack' on 'enemy' territory. There are a number of cases but the others should remain unchanged with this fix since its only adding a 'check' for skipping only when 'allied'.

**Comment by DanVanAtta** [comment]
Ok, then splitting hairs, for logic changing PRs, we have typically very often included new or updated test cases. Why would the lack of tests in other areas excuse the lack of test here?

IF the save game doesn't cover all the test cases, then you're confirming this is not thoroughly tested.

I'm most concerned by the fact that bugs tend to have high locality, chances are very good there are more bugs in this method block. Beyond that, the reason for this fix being correct, or why there was a bug here, is not obvious. A regression test of any kind will help avoid further regressions. I suggest spending effort to improve the code, add *any* kind of test at all, even if it is the simplest of happy test cases.

**Comment by ron-murhammer** [comment]
@DanVanAtta The point is whether we require unit tests for PRs as otherwise this should have already been merged. I'm not arguing whether it would be beneficial. My point is it will take time and I could use the limited time I have more productively elsewhere (reviewing PRs, tackling 1 of 57 open issues, addressing 1 of ~20 TODO AI features, refactoring code, etc). We have 15 post discussion on a 1 line change that only impacts AI code...

Well to be fair there are essentially infinite number of test cases for this method. The key is finding a representative set. Honestly, if you want to debate if something is thoroughly tested look at many of your recent map download PRs where there are obvious issues where I spend 1 min launching the game and find.

There may be more bugs in the method but there is also a good chance the test cases I create don't even find them. I've explained what the bug was and a save showing what it addresses. Writing a unit test that just walks through lines of code isn't really useful and just wastes time to create it and additional code to maintain. Most bugs are also found on the 'unhappy path' and this is an example of that. I could have had a unit test that walks through the method for the most standard case and it work fine. If you are going to unit test something its better to do it properly and come up with a representative set that tests various flows and inputs. 

I'm fine with you mentioning that it would be nice to add unit tests but I don't think its a reason to not merge PRs given the current state of the project and limited time we have to contribute.

**Comment by DanVanAtta** [comment]
I appreciate the well written / thoughtful response. My problem is there is a pattern where we just keep hacking small fixes in this code, I point out various places where the quality is extremely low, and then get a mass of excuses and argument about how that is blowing up the scope, should be done one day, but that day never arrives.

If you'd like a suite of tests, how are we going to get there? If you just build a suite of tests, I would actually suggest that there is low value there. In which case we're almost resigned to no testing in the AI code, and we live with a ball of mud. May as well freeze the project since we won't be making any significant further progress.

So, I really *do* want you to chip away at the code. RAther than typing out the many responses, you coudl have spent 15 minutes doing something to show that the happy case is going to work. Unit tests are mostly regression tests, that is where their value is. For us we get design benefits, since this code was so hard to test, it reveals the massive amount of coupling, the mixing of abstraction, and the tightly coupled dependencies. 

Speaking of limited time, I assume it took you some time to figure out what the right fix was going to be. You did probably a good bit of testing, and work to understand this section of code. By not then taking advantage of that context, and literally taking 10 minutes to do something resembling adding a test, implementing things the right away, reducing copy paste, we just add one more hack. What is the plan to break out of that cycle? Bottom  line, we need to start chipping away at the problem wherever we can.

**Comment by DanVanAtta** [comment]
>  I could have had a unit test that walks through the method for the most standard case and it work fine. If you are going to unit test something its better to do it properly and come up with a representative set that tests various flows and inputs.

I actually disagree here. Thoroughly testing is not always as valuable as something that does a good test. One case can be very brittle and get you stuck in test update hell. Unit tests are effectively the same as regression tests.

There is also a school of thought that says you don't touch code unless there is a failing test. That also any proper bug fix is first exposed with a test, then fixed. Again, we need to start chipping away at this.

One last point, a test that simply walks through the code, is for the most a regression test. There is value in having that.

**Comment by DanVanAtta** [comment]
It could be argued pretty reasonably that this is a hack, not a bug fix. If we just keep hacking the AI code until it "works", I'm not sure what there is really to review. Also, unit tests generally are not written to find bugs, they are most useful to guide code design, and to then serve as regression tests. What is a regression test other than something that walks through the code? How else can you easily and repeatedly prove that this code is working as intended? 

Second, with respect to allocation of time. You chose to work on this patch, and so it must be important. Yet, it's not important enough to add a regression test, or do a manual inspection of nearby code looking for other problems, nor to take advantage of the context switch you incurred to reduce the cyclomatic complexity or the code redundancy. Rather, all that is left for another day, for another developer (perhaps your future self) to spend the same amount of time to understand this code again, when you could have saved that trouble by spending an extra 10 minutes now. From the other perspective, what's the point of reviewing if the majority of suggestions are ignored. Please do not treat only your own time as valuable. With that, I'm also disappointed that there is a pattern where you respond to PR comments with more text, more discussion. I'm frankly not interested in abstract PR discussion. It's generally without value. Even if you disagree with the comments, they were made for a reason, and third parties who come along *will* have the same thoughts. Thus doing something to the code to address those comments in some way, is appropriate. It's not appropriate to respond with explanations etc, because that will only help me now at this point in time, but sometone else will not get that benefit, and they'll see the same code agian, and they'll have the same comments. So explaining stuff in a PR is of almost no value. So please stop spending energy on text, respond to PR comments with code changes until a review would now longer make those comments just based on the code.

Third and final point I wish to make. You must not allow yourself to be in a position where you are such a bottleneck that it's so important for you to rush off to do other things. This has been a severe problem in this project where too much is funnelled to exactly one or two people. It's an example of a broken process with too much overhead. With respect to all those things to do, there are no PRs ready for review right now, every single issue is labelled and chilling. So then it's either work on other AI features, or spend time to make this a bug fix, rather than (arguably) a hack. I'm further most frustrated that I allowed the prior PR to be merged even though technical debt, copy paste code, was added and made an existing problem worse. Similar here, we haven't really moved the needle in terms of fixing other bugs that would be similar to this, nor preventing this in the future, nor reducing technical debt.

**Comment by DanVanAtta** [comment]
@ron-murhammer   " 1 of ~20 TODO AI features,"   Please create some issues around these as feature requests so the rest of the team can see the direction you are heading in.

**Pull request merged**